x-pack/plugins/security/server/routes/users/change_password.ts

```diff
@@ -27,54 +27,65 @@ export function defineChangeUserPasswordRoutes({
-      const providerToLoginWith = config.authc.providers.includes('token') ? 'token' : 'basic';
+      const currentUser = authc.getCurrentUser(request);
+      const isUserChangingOwnPassword =
+        currentUser && currentUser.username === username && canUserChangePassword(currentUser);
+      const currentSession = isUserChangingOwnPassword ? await authc.getSessionInfo(request) : null;
 
-      // If user tries to change own password, let's check if old password is valid first by trying
-      // to login.
-      if (isCurrentUser) {
-        try {
-          const authenticationResult = await authc.login(request, {
-            provider: providerToLoginWith,
-            value: { username, password },
-            // We shouldn't alter authentication state just yet.
-            stateless: true,
-          });
-
-          if (!authenticationResult.succeeded()) {
-            return response.forbidden({ body: authenticationResult.error });
-          }
-        } catch (error) {
-          return response.customError(wrapIntoCustomErrorResponse(error));
-        }
-      }
+      // If user is changing their own password they should provide a proof of knowledge their
+      // current password via sending it in `Authorization: Basic base64(username:current password)`
+      // HTTP header no matter how they logged in to Kibana.
+      const scopedClusterClient = clusterClient.asScoped(
+        isUserChangingOwnPassword
+          ? {
+              headers: {
+                ...request.headers,
+                authorization: `Basic ${Buffer.from(`${username}:${currentPassword}`).toString(
+                  'base64'
+                )}`,
+              },
+            }
+          : request
+      );
 
       try {
-        await clusterClient.asScoped(request).callAsCurrentUser('shield.changePassword', {
+        await scopedClusterClient.callAsCurrentUser('shield.changePassword', {
           username,
           body: { password: newPassword },
         });
+      } catch (error) {
+        // This may happen only if user's credentials are rejected meaning that current password
+        // isn't correct.
+        if (isUserChangingOwnPassword && getErrorStatusCode(error) === 401) {
+          return response.forbidden({ body: error });
+        }
+
+        return response.customError(wrapIntoCustomErrorResponse(error));
+      }
 
-        // Now we authenticate user with the new password again updating current session if any.
-        if (isCurrentUser) {
+      // If user previously had an active session and changed their own password we should re-login
+      // user with the new password and update session. We check this since it's possible to update
+      // password even if user is authenticated via HTTP headers and hence doesn't have an active
+      // session and in such cases we shouldn't create a new one.
+      if (isUserChangingOwnPassword && currentSession) {
```

Comment:
__note:__ we didn't handle the case without session previously, but I think it was a bug. And now we have a proper API that can tell us if we have an active session or not.